lib/MetadataStore.js

```diff
@@ -0,0 +1,385 @@
+   * indices if necessary
+   *
+   * Returns a promise that is resolved upon success, or rejected if an exception occurs
+   */
+  asyncConnect: Task.async(function*() {
+    if (this._conn) {
+      return;
+    }
+
+    try {
+      this._conn = yield Sqlite.openConnection({path: this._path});
+      yield this._conn.execute("PRAGMA journal_mode = WAL;");
+      yield this._conn.execute("PRAGMA foreign_keys = ON;");
+      yield this._createTableSchema();
+    } catch (e) {
+      Cu.reportError("MetadataStore failed to create connection: " + e.message);
+      throw e;
+    }
+  }),
+
+  asyncClose: Task.async(function*() {
+    if (this._conn) {
+      yield this._conn.close();
+      this._conn = null;
+    }
+  }),
+
+  _getMetadataParameters(aRow) {
+    return {
+      cache_key: aRow.cache_key,
+      canonical_url: aRow.canonical_url,
+      title: aRow.title,
+      type: aRow.type || null,
+      description: aRow.description,
+      media_url: aRow.media.url || null,
+      expired_at: aRow.expired_at
+    };
+  },
+
+  _getFaviconParameters(aRow) {
+    return {
+      url: aRow.favicon_url,
+      type: IMAGE_TYPES.favicon,
+      height: 0,
+      width: 0,
+      color: this._rgbToHex(aRow.favicon_colors[0].color)
+    };
+  },
+
+  _getImageParameters(aRow) {
+    return {
+      url: aRow.url,
+      type: IMAGE_TYPES.preview,
+      height: aRow.height,
+      width: aRow.width,
+      color: this._rgbToHex(aRow.colors[0].color)
+    };
+  },
+
+  /**
+  * Convert a RGB array to the Hex form, e.g. [10, 1, 2] => #0A0102
+  */
+  _rgbToHex(rgb) {
+    return "#" + ((1 << 24) + (rgb[0] << 16) + (rgb[1] << 8) + rgb[2]).toString(16).slice(1).toUpperCase();
+  },
+
+  _getLastInsertRowID() {
+    return Task.spawn(function*() {
+      let result = yield this._conn.execute(SQL_LAST_INSERT_ROWID);
+      return result[0].getResultByName("lastInsertRowID");
+    }.bind(this));
+  },
+
+  _getImageIDByURL(url) {
+    return Task.spawn(function*() {
+      let result = yield this._conn.executeCached(SQL_SELECT_IMAGE_ID, [url]);
+      return result[0].getResultByName("id");
+    }.bind(this));
+  },
+
+  /**
+   * Inserts the metadata into the database. It consists of two tables,
+   * i.e. page_metadata for the regular meta information of the page, and
+   * page_images for the favicon and preview images.
+   *
+   * aRows, an array of metadata objects that currently generated by embed.ly
+   * Returns a promise that is resolved upon success, or rejected if an exception occurs
+   */
+  asyncInsert: Task.async(function*(aRows) {
+    const principal = Services.scriptSecurityManager.getSystemPrincipal();
+    for (let row of aRows) {
+      yield this._conn.executeTransaction(function*() {
+        let metadata_id;
+        let image_ids = [];
+
+        /* attach favicon to canonical_url for Places */
+        PlacesUtils.favicons.setAndFetchFaviconForPage(
+          NetUtil.newURI(row.canonical_url),
```

Comment:
Good catch, it's a bug. The `canonical_url` should be obtained from the embed.ly response payload for now.

@jaredkerim can you confirm that which field in the current embed.ly payload could be used as the `canonical_url`? `provider_url`, `original_url`, or `url`?

I recall you mentioned aobut `provider_url` last week, but it does not look like that way to me.